packages/zoe/src/contracts/vpool-xyk-amm/multipoolMarketMaker.js

```diff
@@ -0,0 +1,193 @@
+ * tokens can be exchanged with each other, but only through the central
+ * token. For example, if X and Y are two token types and C is the central
+ * token, a swap giving X and wanting Y would first use the pool (X, C) then the
+ * pool (Y, C). There are no liquidity pools directly between two secondary
+ * tokens.
+ *
+ * There should only need to be one instance of this contract, so liquidity can
+ * be shared as much as possible.
+ *
+ * When the contract is instantiated, the central token is specified in the
+ * terms. Separate invitations are available by calling methods on the
+ * publicFacet for adding and removing liquidity and for making trades. Other
+ * publicFacet operations support querying prices and the sizes of pools. New
+ * Pools can be created with addPool().
+ *
+ * When making trades or requesting prices, the caller must specify either a
+ * maximum input amount (swapIn, getInputPrice) or a minimum output amount
+ * (swapOut, getOutPutPrice) or both. For swaps, the required keywords are `In`
+ * for the trader's `give` amount, and `Out` for the trader's `want` amount.
+ * getInputPrice and getOutputPrice each take two Amounts. The price functions
+ * return both amountIn (which may be lower than the original amount) and
+ * amountOut (which may be higher). When both prices are specified, no swap will
+ * be made (and no price provided) if both restrictions can't be honored.
+ *
+ * When adding and removing liquidity, the keywords are Central, Secondary, and
+ * Liquidity. adding liquidity has Central and Secondary in the `give` section,
+ * while removing liquidity has `want` and `give` swapped.
+ *
+ * Transactions that don't require an invitation include addPool and the
+ * queries: getInputPrice, getOutputPrice, getPoolAllocation,
+ * getLiquidityIssuer, and getLiquiditySupply.
+ *
+ * @type {ContractStartFn}
+ */
+const start = zcf => {
+  // This contract must have a "Central" keyword and issuer in the
+  // IssuerKeywordRecord.
+  // poolFee is the portion of the fees that go into the pool (in BP).
+  // protocolFee is the portion of the fees that are shared with validators
+  const {
+    brands: { Central: centralBrand },
+    timer,
+    poolFee,
+    protocolFee,
+  } = zcf.getTerms();
+  assertIssuerKeywords(zcf, ['Central']);
+  assert(centralBrand !== undefined, X`centralBrand must be present`);
+
+  /** @type {WeakStore<Brand,XYKPool>} */
+  const secondaryBrandToPool = makeWeakStore('secondaryBrand');
+  const getPool = secondaryBrandToPool.get;
+  const initPool = secondaryBrandToPool.init;
+  const isSecondary = secondaryBrandToPool.has;
+
+  const quoteIssuerKit = makeIssuerKit('Quote', AssetKind.SET);
+
+  // For now, this seat collects protocol fees. It needs to be connected to
+  // something that will extract the fees.
+  const { zcfSeat: protocolSeat } = zcf.makeEmptySeatKit();
+
+  const getLiquiditySupply = brand => getPool(brand).getLiquiditySupply();
```

Comment:
_IOU a separate issue on this out-of-scope comment_

I wondered why `tsc` doesn't complain about the case where `getPool()` returns `undefined`, but I see that `store.get()` throws on key not found.

The standard js [Map.get](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get) returns `undefined` on key not found. Is it wise to use the same name with different semantics?

I see that we use `init` as distinct from `set` to mean "throw if it's already there."